### ISSUE TITLE
Unable to write MBTiles with Gdalwarp

### ISSUE THREAD

**Issue by eddy-geek** [OSGeo/gdal#3460]
Zoom levels are not determined from `-tr`, making it mandatory to go through an extra step (eg tif) when reprojecting a geotif to mbtile.

Sometimes the extra step is not practical due to file size/quality loss.

## Expected behavior and actual behavior.

With GeoPkg I can do `gdalwarp -t_srs EPSG:3857 *.tif output.gpkg -tr 2.388657 2.388657` (as said in https://github.com/OSGeo/gdal/issues/1160)

But if I just replace `gpkg` with `mbtiles` in the command above, I get

> ERROR 1: Cannot find min and max zoom_level
> ERROR 6: Could not find an appropriate zoom level that matches raster pixel size

## Steps to reproduce the problem.

With a raster taken from another issue which seems to reproduce the issue:

```bash
gdal_create -outsize 16578 21503 -a_srs EPSG:2154 -co compress=deflate -a_ullr -10071418.950 3656755.453 -10071169.382 3656431.744 mbtest.tif

gdalwarp -t_srs EPSG:3857 *.tif output.mbtiles -tr 2.388657 2.388657

gdalwarp -t_srs EPSG:3857 *.tif output.gpkg -tr 2.388657 2.388657                                                                   ⏎
```

> Creating output file that is 107P x 105L
> ERROR 6: Could not find an appropriate zoom level that matches raster pixel size

> Creating output file that is 107P x 105L.
>        - 'VirtualXPath'        [XML Path Language - XPath]
> Processing mbtest.tif [1/1] : 0...10...20...30...40...50...60...70...80...90...100 - done.

## Operating system
Ubuntu 20.04 64 bit

## GDAL version and provenance

```
gdalinfo --version
GDAL 3.2.1, released 2020/12/29
```
(ubuntugis-unstable PPA)

**Comment by jratike80** [comment]
Have you tried to use VRT as an outputformat? If it works you can avoid problems with size/quality loss.

**Comment by eddy-geek** [comment]
I agree there may be good work-arounds, but I believe that with a bit more code-sharing between gpkg and mbtiles backend, no work-around should be needed ;-)

`-tr` should work, and it would be even easier if it were computed from the `MINZOOM`/`MAXZOOM` arguments (currently reserved to the vector MBTiles driver):

ZOOM 15 -> `-tr 4.777314267823516 -...`
ZOOM 16 -> `-tr 2.388657133911758 -...`

In the same vein, even something as simple as `gdal_create -outsize 3072 5630 -a_srs EPSG:3857 -a_ullr 626172 5831133 860946 5400829 -bands 3 -burn 255 white.mbtiles` fails with the same error.

**Comment by EvertEt** [comment]
@eddy-geek How did you calculate the tr values per zoom level? Are these valid everywhere or depending on location?
I am trying to understand if this issue is linked to https://github.com/OSGeo/gdal/issues/3126.

**Comment by jratike80** [comment]
@eddy-geek I write down the workarounds because they are needed until somebody improves gdalwarp either on their spare time or contracted.

Your comparison with GeoPackage is not valid. If you read https://gdal.org/drivers/raster/gpkg.html you'll see 

> Tiling schemes
> By default, conversion to GeoPackage will create a custom tiling scheme, such that the input dataset can be losslessly converted, both at the pixel and georeferencing level (if using a lossless tile format such as PNG). That tiling scheme is such that its origin (min_x, max_y) in the gpkg_tile_matrix_set table perfectly matches the top left corner of the dataset, and the selected resolution (pixel_x_size, pixel_y_size) at the computed maximum zoom_level of the gpkg_tile_matrix table will match the pixel width and height of the raster.

For a proper test you must use the same tiling scheme:

```
gdalwarp -of gpkg -t_srs epsg:3857 -co tiling_scheme=GoogleMapsCompatible -tr 1.2 1.2 p4433h.tif gp.gpkg
Creating output file that is 11085P x 11100L.
ERROR 6: Could not find an appropriate zoom level of GoogleMapsCompatible tiling scheme that matches raster pixel size
ERROR 1: Raster table gp not correctly initialized due to missing call to SetGeoTransform()
```

So code sharing between GPKG and MBTiles will not help with this issue.

@EvertEt The Google-style Web Mercator tiles have fixed nominal pixel sizes all over the world https://wiki.openstreetmap.org/wiki/Zoom_levels. I say nominal because the true pixel size on the ground varies.

There is not so direct connection with #3126 that one one fix would cure them both.

**Comment by eddy-geek** [comment]
@ EvertEt : @jratike80 answered better that I could ;-)

@jratike80 : as for the workaround, I have 3 files A / B / C having different projections (because they are slopes `color-relief`-ed from different DEMs) -- so I get:
```
gdalbuildvrt does not support heterogeneous projection: expected RGF93 / Lambert-93, got ED50 / UTM zone 32N. Skipping B.tif
```

So my current workaround is:
```
gdalwarp -t_srs EPSG:3857 -tr 2.388657133911758 -2.388657133911758 A.tif B.tif C.tif tmp.tif
```

This way, I avoid the quality issue by having only one resampling (at least I hope).
As for disk space, *tmp.tif* is 5.6 GB even with:
`-co tiled=yes -co blockXsize=1024 -co blockYsize=1024 -co bigtiff=yes -co compress=zstd -co predictor=2 -co zstd_level=1`
 whereas the mbtiles is 2.3 GB in JPG-75%, and 4.4 GB in PNG (which surprised me -- maybe it's because *tmp.tif* isn't rectangular, so mbtiles benefits from not writing transparent tiles). 

I could have tmp.tif using jpg but I'm afraid of the double lossy compression, and anyway it's still 2 GB more than needed otherwise ;-)

**Comment by jratike80** [comment]
Deflate compression packs uniform empty area very well, perhaps zstd also but you could have a try. If slope is 16 bit you can also try predictor=3.
